init.go

```diff
@@ -0,0 +1,130 @@
+package surge
+
+// This file implements the centralized initialization framework for surge.
+// The core interface is Initializer. Each file or module that needs to have some
+// initialization logic before calling main should create a structure which includes
+// BaseInitializer structure. The structure should redefine the Initialize() method
+// which should call the function that performs initialization. Each initialization
+// can have a dependency list of Initializers. The dependency should be met before
+// the initializer is called.
+// One example: config should be called before model. So model initialization
+// will have a dependency set for config.
+//
+// With this new framework, no files other than this file will have the init method.
+// All files or modules that need initialization will define their own initializer and
+// set the dependency correctly and then add it to the global initializers list using
+// the AddInitializer() function. This should be called as a variable initialization in
+// each file so that they get called befor the init() function here.
+
+import (
+	"fmt"
+	)
+
+
+type InitializerState int
+
+const (
+	InitializerNotProcessed InitializerState = iota
+	InitializerBlocked
+	InitializerDeadlocked
+	InitializerProcessed
+	InitializerFailed
+)
+
+type Initializer interface  {
+	Initialize() InitializerState
+	GetName() string
+	GetState() InitializerState
+	SetState(InitializerState)
+	BlockedOn(dependency Initializer) bool
+	SetDependency(initializer Initializer)
+	GetDependencies() []Initializer
+}
+
+var initializers = make(map[string]Initializer)
+
+func AddInitializer(initializer Initializer) bool {
+	name := initializer.GetName()
+	assert(initializers[name] == nil)
+	initializers[name] = initializer
+	fmt.Printf("Added Initializer: %s\n", name)
+
+	return true
+}
+
+func initialize(initializer Initializer) InitializerState{
+	state := initializer.Initialize()
+	initializer.SetState(state)
+
+	return state
+}
+
+func init() {
+	fmt.Println("init called from init.go")
+	count := len(initializers)
+	for _, initializer := range initializers {
+		if initializer.GetState() != InitializerNotProcessed {
+			continue
+		}
+		for _, dep := range initializer.GetDependencies() {
+
+			// Check if we will deadlock
+			if dep.BlockedOn(initializer) {
+				initializer.SetState(InitializerDeadlocked)
+				// panic here
+				return 
+			}
+
+			status := initialize(dep)
+			if status == InitializerBlocked {
+				initializer.SetState(InitializerBlocked)
+			}
+		}
+		if initializer.GetState() != InitializerBlocked {
+			initialize(initializer)
+			count--
+		}
+	}
+	assert(count == 0)
+}
+
+type BaseInitializer struct {
+	name	string
```

Comment:
Typically, all the types are declared at the top. I'm also often using some kind of a multi-line delimiter, to separate different types and methods within the same file.

Overall, this code will replace Go's arbitrary init() sequencing with fully supported tree-like initialization dependencies. But my sense is, it's a bit over-engineering at this point. Take a look at the init() calls we have now:

config.go:166:func init() {
m1.1.go:31:func init() {
...
m9.go:29:func init() {
ma.go:43:func init() {
mb.go:35:func init() {
model.go:57:func init() {

What we probably want is to init framework ( [log, model, config].go ) - first, and all the m*.go models - second. That's all. The rest initialization is done today via Configure() interface. 

What I'm saying is, we don't really need generic tree-like dependencies. We simply need two passes - PASS1 to initialize framework, PASS2 to initialize models. Anything beyond these two steps does not belong to init() and must be done later via Configure() or in the first lines of Run() method.